### PR TITLE
feature/hypermedia

### PR DESCRIPTION
**Pull request by Ivan-Marquez**
# Hypermedia
> Adding HEATEOAS to Style Point

## Change log
+ Added centralized route map to improve reusability of routes in express and Hypermedia modules.
+ Added hypermedia module to format API responses with hypermedia links.
+ Added links with relations for the following entities:
    + User
    + Log
+ Added unit tests for hypermedia module.

## Additional changes
+ Module added to handle pagination errors.
+ Deleted id property on all mongoose models.
+ All models where renamed (removed `.server` from name).
+ Changed services to return only a few attributes when responding with a collection.
+ Added url's to manage previous and next page properties in pagination module.
+ Added [slugify](https://www.npmjs.com/package/slugify) package to generate valid url's.
+ Added `authorize` middleware to authorize requests by user's role.
+ Modified and added unit tests to reach 100% code coverage.

### CODE REVIEW

File: src/config/express.js [hunk 1]
```diff
@@ -57,19 +57,14 @@ module.exports = () => {
    */
   router.templateModel = {
     year: new Date().getFullYear(),
-    gitUrl: package.homepage,
-    version: package.version
+    gitUrl: config.homepage,
+    version: config.version
   }
 
-  /**
```
Comment (kevinpr007) [inline src/config/express.js]: We must leave this comment. It's important

File: src/models/location.model.js [hunk 2]
```diff
@@ -52,7 +52,11 @@ const locationSchema = new Schema({
 
 locationSchema.set('toJSON', {
   getters: true,
-  virtuals: true
+  virtuals: true,
+  transform: (doc, ret) => {
+    delete ret.id
```
Comment (kevinpr007) [inline src/models/location.model.js]: It's the mongoose Id?

File: src/models/roles.enum.js [hunk 3]
```diff
@@ -0,0 +1,11 @@
+/*
+ sysAdmin:    This is the administrator of the main application. This role is only for the owners of Style Point.
+ Admin:         This is the administrator of one store/location inside the framework.
+ Manager:       Each store/location has a main manager to manage all tasks related to their employees.
+ Employee:      A member of the manager's crew. They will be using the application to manage a normal
+                day schedule in the store.
+ User:          They will be the clients in the system. User role will be the role with the least privileges in the system.
+ */
+module.exports = () => {
+  return ['sysAdmin', 'Admin', 'Manager', 'Employee', 'User']
```
Comment (kevinpr007) [inline src/models/roles.enum.js]: We should create a enum class to use it in this file, controllers, etc

File: src/models/schedule.model.js [hunk 4]
```diff
@@ -1,7 +1,7 @@
 const mongoose = require('mongoose')
 const Schema = mongoose.Schema
 const mongoDB = require('../config/mongoose.collections.json')
-const types = require('./schedule-types.server.enum')()
+const types = require('./schedule-types.enum.js')()
```
Comment (kevinpr007) [inline src/models/schedule.model.js]: Remove this .js

File: src/models/user.model.js [hunk 5]
```diff
@@ -2,7 +2,7 @@ const mongoose = require('mongoose')
 const Schema = mongoose.Schema
 const bcrypt = require('bcrypt-nodejs')
 const mongoDB = require('../config/mongoose.collections.json')
-const roles = require('./roles.server.enum')()
+const roles = require('./roles.enum.js')()
```
Comment (kevinpr007) [inline src/models/user.model.js]: Remove all .js extensions

File: src/routes/log.routes.js [hunk 6]
```diff
@@ -1,8 +1,11 @@
...
 const logController = require('../controllers/log.controller')(logService)
+const authenticate = require('passport').authenticate('jwt', {session: false})
+const authorize = require('../services/auth.service')().authorize
+const routes = require('./routes.config')
 
 module.exports = router => {
-  router.get('/logs', logController.getAllLogs)
-  router.get('/logs/code/:code', logController.getByCode)
-  router.get('/logs/status/:status', logController.getByStatus)
+  router.get(routes.get('getLogs').path, authenticate, authorize(['sysAdmin']), logController.getAllLogs)
```
Comment (kevinpr007) [inline src/routes/log.routes.js]: In this part we can use the enum roles

File: src/services/auth.service.js [hunk 7]
```diff
@@ -40,10 +41,22 @@ let authService = () => {
...
   }
 
+  let authorize = roles => {
+    return (req, res, next) => {
+      if (!roles.includes(req.user.roles)) { //TODO: Check if a user could have multiple roles.
+        let err = new Error('Your user does not have the required role(s) to execute this action.')
+        err.status = HttpStatus.UNAUTHORIZED
+        throw err
+      }
+      next()
```
Comment (kevinpr007) [inline src/services/auth.service.js]: Why we are using this next? This is not a middleware or this other thing.

File: src/models/location.model.js (hunk 2 above)
Comment (Ivan-Marquez) [inline src/models/location.model.js]: From the [documentation](http://mongoosejs.com/docs/guide.html#id):

Mongoose assigns each of your schemas an id virtual getter by default which returns the documents _id field cast to a string, or in the case of ObjectIds, its hexString.

Basically, the one that is required for documents is `_id`. `id` is just a getter.

File: src/services/auth.service.js (hunk 7 above)
Comment (Ivan-Marquez) [inline src/services/auth.service.js]: It *IS* a middleware function. Calling `next()` is required to go to the next middleware function. As you already noticed, we're calling the `authorize()` middleware *before* calling the controller's function.

File: src/models/roles.enum.js (hunk 3 above)
Comment (Ivan-Marquez) [inline src/models/roles.enum.js]: There are no enums in javascript. The only thing we could do to have something similar is an object. Something like:

```js
let roles = {
   sysAdmin: 0,
   Admin: 1
}
```

But it's something we could do later when we have more roles.

File: src/models/location.model.js (hunk 2 above)
Comment (kevinpr007) [inline src/models/location.model.js]: Perfect!

File: src/models/roles.enum.js (hunk 3 above)
Comment (kevinpr007) [inline src/models/roles.enum.js]: I mean we can do something like this:

{
  "Model": {
    "Location": "Location",
    "User": "User",
    "Workshift": "Workshift",
    "Service": "Service",
    "Schedule": "Schedule",
    "Log": "Log"
  },
  "Collection": {
    "Location": "sp_location",
    "User": "sp_user",
    "Workshift": "sp_workshift",
    "Service": "sp_service",
    "Schedule": "sp_schedule",
    "Log": "sp_log"
  }
}

**Pull request merged**